File: frontend/src/services/scoremystreet.ts
/** ScoreMyStreet 评分响应 */
export interface ScoreMyStreetScore {
  postcode: string | null
  report_url: string
  search_url: string
  overall_score: number | null
  safety_score: number | null
  schools_score: number | null
  amenities_score: number | null
  transport_score: number | null
  connectivity_score: number | null
  area_name: string | null
  crime_count: number | null
  crime_rate: string | null
  crime_types: Record<string, number> | null
  postcode_district: string | null
  date_processed: string | null
  fetched: boolean
  source: string
  schools_count: number | null
  schools_info: string | null
  supermarkets_count: number | null
  parks_count: number | null
  gyms_count: number | null
  ev_charging_count: number | null
  nearest_station: string | null
  nearest_station_distance: number | null
  stations_count: number | null
  full_fibre_coverage: string | null
  ultrafast_coverage: string | null
  superfast_coverage: string | null
}

export const scoreMyStreetService = {
  /**
   * 获取地址对应的 ScoreMyStreet 评分
   * @param address 完整地址（应包含 UK 邮编）
   * @param country 国家代码（可选，例如 'GB'）
   */
  async getScore(address: string, country?: string): Promise<ScoreMyStreetScore> {
    const params = new URLSearchParams({ address })
    if (country) params.set('country', country)
    const token = localStorage.getItem('access_token')
    const headers: Record<string, string> = { 'Content-Type': 'application/json' }
    if (token) headers['Authorization'] = `Bearer ${token}`
    
    const response = await fetch(`/api/v1/scoremystreet/score?${params.toString()}`, {
      method: 'GET',
      headers,
      signal: AbortSignal.timeout(30000),
    })
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }
    
    return response.json()
  },
}

File: backend/app/api/v1/routes/scoremystreet.py
"""ScoreMyStreet 评分路由 - 根据 UK 地址查询其在 scoremystreet.com 的评分"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query

from app.services.scoremystreet_service import (
    get_scoremystreet_score,
    is_uk_address,
)

router = APIRouter()


@router.get("/score")
async def get_score(
    address: str = Query(..., min_length=1, max_length=500),
    country: str | None = Query(default=None, max_length=8),
) -> dict:
    """根据地址获取 ScoreMyStreet 评分信息

    返回的字段：
    - postcode: 邮编
    - report_url: 报告页 URL
    - search_url: 搜索页 URL
    - overall_score: 0-100 的综合评分
    - safety_score: 安全评分（0-5）
    - schools_score: 学校评分（0-100）
    - amenities_score: 设施评分（0-100）
    - transport_score: 交通评分（0-100）
    - connectivity_score: 网络评分（0-100）
    - area_name: 区域名称
    - crime_count: 近6个月犯罪数量
    - fetched: 是否成功从网站抓取并解析到评分
    - source: 数据来源
    """
    if not is_uk_address(address, country):
        raise HTTPException(
            status_code=400,
            detail="ScoreMyStreet 仅支持英国地址，请确认地址属于 UK",
        )

    result = await get_scoremystreet_score(address)
    return result.to_dict()
